File: src/main/java/net/artux/pdanetwork/service/user/ban/BanServiceImpl.java
package net.artux.pdanetwork.service.user.ban;

import lombok.RequiredArgsConstructor;
import net.artux.pdanetwork.entity.user.BanEntity;
import net.artux.pdanetwork.entity.user.UserEntity;
import net.artux.pdanetwork.models.user.ban.BanDto;
import net.artux.pdanetwork.models.user.ban.BanMapper;
import net.artux.pdanetwork.repository.user.BanRepository;
import net.artux.pdanetwork.service.user.UserService;
import net.artux.pdanetwork.utills.security.IsModerator;
import org.slf4j.Logger;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;

@Service
@RequiredArgsConstructor
public class BanServiceImpl implements BanService {

    private static final ConcurrentHashMap<UUID, BanEntity> currentBanMap = new ConcurrentHashMap<>();
    private final Timer timer = new Timer();

    private final Logger logger;
    private final BanMapper banMapper;
    private final BanRepository repository;
    private final UserService userService;

    @Override
    public List<BanDto> getBans(UUID userId) {
        return banMapper.dto(repository.findAllByUserId(userId));
    }

    @Override
    public boolean isBanned(UUID userId) {
        return currentBanMap.containsKey(userId);
    }

    @Override
    public boolean setChatBan(UUID userId) {
        return userService.setChatBan(userId);
    }

    @Override
    public BanDto applySystemBan(UUID userId, String reason, String message, int secs) {
        BanEntity banEntity = new BanEntity();
        banEntity.setBy(null);
        return banUser(userId, reason, message, secs, banEntity);
    }

    private BanDto banUser(UUID userId, String reason, String message, int secs, BanEntity banEntity) {
        UserEntity user = userService.getUserById(userId);
        banEntity.setUser(user);
        banEntity.setReason(reason);
        banEntity.setMessage(message);
        banEntity.setSeconds(secs);

        currentBanMap.put(userId, repository.save(banEntity));
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                currentBanMap.remove(userId);
            }
        }, 1000L * secs);
        logger.info("User {} banned for {} seconds, reason: ", user.getLogin(), secs, reason);
        return getCurrentBan(userId);
    }

    @Override
    @IsModerator
    public BanDto applyBan(UUID userId, String reason, String message, int secs) {
        UserEntity user = userService.getUserById();
        BanEntity banEntity = new BanEntity();
        banEntity.setBy(user);
        return banUser(userId, reason, message, secs, banEntity);
    }

    @Override
    public BanDto getCurrentBan(UUID userId) {
        return banMapper.dto(currentBanMap.get(userId));
    }

    @Override
    @Transactional
    public boolean clearAllBans(UUID uuid) {
        currentBanMap.remove(uuid);
        repository.deleteAllByUserId(uuid);
        return true;
    }

    @Override
    @Transactional
    public boolean clearBan(UUID banId) {
        BanEntity ban = repository.findById(banId).orElseThrow(() -> new RuntimeException("Ban not found"));
        currentBanMap.remove(ban.getUser().getId());
        repository.deleteById(banId);
        return true;
    }

    @Override
    public Map<UUID, BanDto> getCurrentBans() {
        Map<UUID, BanDto> bans = new HashMap<>();
        currentBanMap.forEach((uuid, banEntity) -> {
            bans.put(uuid, banMapper.dto(banEntity));
        });
        return bans;
    }

}

File: pdanet/src/main/java/net/artux/pdanetwork/service/user/ban/BanServiceImpl.kt
package net.artux.pdanetwork.service.user.ban

import lombok.RequiredArgsConstructor
import mu.KLogging
import net.artux.pdanetwork.entity.mappers.BanMapper
import net.artux.pdanetwork.entity.user.BanEntity
import net.artux.pdanetwork.models.user.ban.BanDto
import net.artux.pdanetwork.repository.user.BanRepository
import net.artux.pdanetwork.service.user.UserService
import net.artux.pdanetwork.utils.security.ModeratorAccess
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Duration
import java.util.Timer
import java.util.TimerTask
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

@Service
@RequiredArgsConstructor
open class BanServiceImpl(
    private val banMapper: BanMapper,
    private val repository: BanRepository,
    private val userService: UserService,
) : BanService {

    private val timer = Timer()

    override fun getBans(userId: UUID): List<BanDto> {
        return banMapper.dto(repository.findAllByUserId(userId))
    }

    override fun isBanned(userId: UUID): Boolean {
        return currentBanMap.containsKey(userId)
    }

    override fun setChatBan(userId: UUID): Boolean {
        return userService.setChatBan(userId)
    }

    override fun applySystemBan(userId: UUID, reason: String, message: String, secs: Long): BanDto {
        val banEntity = BanEntity()
        banEntity.by = null
        return banUser(userId, reason, message, secs, banEntity)
    }

    private fun banUser(userId: UUID, reason: String, message: String, secs: Long, banEntity: BanEntity): BanDto {
        val user = userService.getCurrentUser(userId)
        banEntity.user = user
        banEntity.reason = reason
        banEntity.message = message
        banEntity.seconds = secs.toInt()
        currentBanMap[userId] = repository.save(banEntity)
        timer.schedule(
            object : TimerTask() {
                override fun run() {
                    currentBanMap.remove(userId)
                }
            },
            Duration.ofSeconds(secs).toMillis()
        )
        logger.info("User ${user.login} banned for $secs seconds, reason: $reason")
        return getCurrentBan(userId)
    }

    @ModeratorAccess
    override fun applyBan(userId: UUID, reason: String, message: String, secs: Long): BanDto {
        val user = userService.getCurrentUser()
        val banEntity = BanEntity()
        banEntity.by = user
        return banUser(userId, reason, message, secs, banEntity)
    }

    override fun getCurrentBan(userId: UUID): BanDto {
        return banMapper.dto(currentBanMap[userId])
    }

    @Transactional
    override fun clearAllBans(uuid: UUID): Boolean {
        currentBanMap.remove(uuid)
        repository.deleteAllByUserId(uuid)
        return true
    }

    @Transactional
    override fun clearBan(banId: UUID): Boolean {
        val ban = repository.findById(banId).orElseThrow { RuntimeException("Ban not found") }
        currentBanMap.remove(ban.user.id)
        repository.deleteById(banId)
        return true
    }

    override fun getCurrentBans(): Map<UUID, BanDto> {
        val bans: MutableMap<UUID, BanDto> = HashMap()
        currentBanMap.forEach { (uuid: UUID, banEntity: BanEntity?) -> bans[uuid] = banMapper.dto(banEntity) }
        return bans
    }

    companion object : KLogging() {
        private val currentBanMap = ConcurrentHashMap<UUID, BanEntity>()
    }
}
